[sunspec] Modbus: SunSpec bundle basic version initial contribution

mrbig opened the pull request:
SunSpec is an open standard for solar inverters and other related devices to share data about their internal state. The standard is implemented by several vendors like ABB, Fronius, SMA, Schneider Electric, Solaredge, etc.

The goal of this work is to add user friendly support of these devices to openHAB.

The standard is built on the Modbus protocol, so this work is heavily based on the Modbus binding. Also the Bluetooth binding and several of it's solutions were taken as an example how to extend an already existing binding with a new bundle.

Related issue: #3216

This is a stripped down version of a SunSpec bundle aimed to ease reviewing as requested by @davidgraeff 

Original PR with all the features can be found here: #4220 

This version contains only a limited support for single phase inverters, but no auto discovery or any other fancy features are included.

Other changes from my original PR:
 - migrated to the new build system
 - using NonNull values wherever possible
 - modell classes were moved to a dto package and highly simplified
 - reflected code changes in the modbus package
 - other minor code changes

wborn commented on bundles/org.openhab.binding.modbus.sunspec/src/main/resources/ESH-INF/config/config-descriptions.xml:
```diff
@@ -0,0 +1,41 @@
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:config-description="http://eclipse.org/smarthome/schemas/config-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/config-description/v1.0.0
+        http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
```
The incremental build seems to work now. :+1: 

You can fix the errors by updating the scheme URLs from `http://eclipse.org/smarthome/schemas` to  `https://openhab.org/schemas`

mstormi commented:
any particular reason why this isn't merged yet or was it just forgotten about ?

Hilbrand commented:
It's not merged because it's not reviewed because there is a backlog.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/README.md:
```diff
@@ -0,0 +1,268 @@
+
+This bundle is an extension for the Modbus binding to support the SunSpec protocol.
+
+SunSpec is a format for inverters and smart meters to communicate over the Modbus protocol. It defines how common parameters like AC/DC voltage and current, lifetime produced energy, device temperature etc can be read from the device.
```
fixed this.

mrbig commented on buildci.sh:
```diff
@@ -50,8 +50,8 @@ fi
 if [[ ! -z "$CHANGED_DIR" ]] && [[ -e "bundles/$CHANGED_DIR" ]]; then
     echo "Single addon pull request: Building $CHANGED_DIR"
     echo "MAVEN_OPTS='-Xms1g -Xmx2g -Dorg.slf4j.simpleLogger.log.org.openhab.tools.analysis.report.ReportUtility=DEBUG -Dorg.slf4j.simpleLogger.defaultLogLevel=WARN'" > ~/.mavenrc
-    cd "bundles/$CHANGED_DIR"
```
Thanks. Rebased my branch so this wasn't necessary anymore.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/README.md:
```diff
@@ -0,0 +1,268 @@
+
+# IMPORTANT: under merge
+
+** IMPORTANT: this version of this bundle is being merged into OpenHAB. This will be done in small steps - this means that not everything in this readme is supported at the moment **
```
Fixed this, and all other occurences I could find.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/README.md:
```diff
@@ -0,0 +1,268 @@
+
+This group contains general operational information about the device.
+
+| Channel Type ID      | Item Type | Description                                                         |
```
Thanks, changed.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/pom.xml:
```diff
@@ -0,0 +1,31 @@
+	<parent>
+		<groupId>org.openhab.addons.bundles</groupId>
+		<artifactId>org.openhab.addons.reactor.bundles</artifactId>
+		<version>2.5.0-SNAPSHOT</version>
```
Thanks, updated to the new version.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/pom.xml:
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+	<modelVersion>4.0.0</modelVersion>
```
Thanks for pointing out, changed the indentation.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecConfiguration.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
```
Thank you, updated the licenses

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/InverterHandler.java:
```diff
@@ -0,0 +1,106 @@
+        super(thing, managerRef);
+    }
+
+    /**
```
Thanks, fixed the missing javadoc.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/parser/SunspecParser.java:
```diff
@@ -0,0 +1,42 @@
+     * This method should parser an incoming register array and
+     * return a not-null sunspec block
+     */
+    public T parse(ModbusRegisterArray raw);
```
Removed, thanks for noticing

mrbig commented on bundles/pom.xml:
```diff
@@ -200,6 +200,7 @@
     <module>org.openhab.binding.sonyprojector</module>
     <module>org.openhab.binding.spotify</module>
     <module>org.openhab.binding.squeezebox</module>
+    <module>org.openhab.binding.modbus.sunspec</module>
```
I've moved it to the right place.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,452 @@
+    }
+
+    /**
+     * Handle incoming commands. This binding is read-only be default
```
Fixed the typo, thanks.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,452 @@
+    protected Supplier<ModbusManager> managerRef;
+
+    /**
+     * Instances of this handler should get a reference to the modbus managet
```
Fixed the typo, thanks.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecConfiguration.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    private int length;
+
+    public Integer getAddress() {
```
Thank you, replaced it with int.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecConfiguration.java:
```diff
@@ -0,0 +1,79 @@
+        this.address = address;
+    }
+
+    public Integer getLength() {
```
Replaced this as well with int.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecConfiguration.java:
```diff
@@ -0,0 +1,79 @@
+     * Sets refresh period in milliseconds
+     */
+    public void setRefreshMillis(long refresh) {
+        this.refresh = refresh / 1000;
```
This wasn't used at all, so I removed the setter.

We set the raw value (seconds), but I've put there the utility getter to hide the conversion.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/InverterHandler.java:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * Parser used to convert incoming raw messages into model blocks
+     */
+    private InverterModelParser parser = new InverterModelParser();
```
Thank you, marked it as final

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecHandlerFactory.java:
```diff
@@ -0,0 +1,102 @@
+     * @param manager the modbus manager from org.openhab.io.transport.modbus package
+     */
+    @Reference(service = ModbusManager.class, cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.STATIC, unbind = "unsetManager")
+    public void setManager(ModbusManager manager) {
```
Thanks for pointing out, changed this and that made many things simpler.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/resources/ESH-INF/config/config-descriptions.xml:
```diff
@@ -0,0 +1,39 @@
+	<config-description uri="thing-type:sunspec:modbusconfig">
+
+		<parameter name="refresh" type="integer" min="0">
+			<label>interval (s)</label>
```
Sorry, didn't know about that. Changed wherever I could find.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/resources/ESH-INF/thing/inverter-channel-types.xml:
```diff
@@ -0,0 +1,85 @@
+
+
+	<channel-type id="status-type">
+		<item-type>Number</item-type>
```
Thanks for the guidance, changed, and now the status display works as I would expect.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,452 @@
+    public synchronized void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        logger.debug("bridgeStatusChanged for {}. Reseting handler", this.getThing().getUID());
+        if (bridgeStatusInfo.getStatus() == ThingStatus.ONLINE && getThing().getStatus() == ThingStatus.OFFLINE) {
+            initialize();
```
Thanks, I've refactored how the initialize/dispose/bridgeStatusChanged works. Can I ask you to review it?

thank you

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,452 @@
+        unregisterPollTask();
+        unregisterEndpoint();
+        if (reason.isPresent()) {
+            updateStatus(ThingStatus.OFFLINE, reason.get());
```
Thank you, I've removed this.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,452 @@
+     * Get a reference to the modbus endpoint
+     */
+    @SuppressWarnings("null")
+    private synchronized void connectEndpoint() {
```
Thanks, I've rethought that method, and I think it was unnecessary to mark it synchronized.

mrbig commented:
> Thanks for splitting up. I've reviewed the code. 1 general comment. Did you consider using UoM (or QuantityType) channels? 

Thanks @Hilbrand for reviewing these. I didn't know about UoM but now I've added those also. There are 2 units missing from the core, so I've created a [PR for those](https://github.com/openhab/openhab-core/pull/1347).

I believe I've fixed all you request now.  I you would like to change anything else please let me know.

mrbig commented:
@Hilbrand  One more question regarding UoM: should my binding convert big values to a more appropriate metric prefix, or is this something the user should be able to do in the UI?

For example: lifetime energy can be high like 15200108Wh. Should I emit such a value, or is it more appropriate to first scale it to something like 15.2MWh?

(Paper UI did not scale down for me at the moment)

aligot-cblue commented on bundles/org.openhab.binding.modbus.sunspec/src/main/resources/ESH-INF/thing/inverter-types.xml:
```diff
@@ -0,0 +1,36 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="inverter-single-phase">
```
Do you plan to support three-phase inverter ? I can provider modbus/tcp access to mine if needed.

Best regards,

a.

mrbig replied:
> Do you plan to support three-phase inverter ? I can provider modbus/tcp access to mine if needed.

Thanks, yes, I already had implemented it in the full version, but then I have split it up into smaller PR-s for easier review. This PR contains only the most basic functionality and focuses on the skeleton of the bundle.

2/3 phase inverters are then almost just repetitions of the same components. 

After hopefully this bundle gets accepted I'll submit the autodiscovery feature, after those I can start add more device specific functions.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/README.md:
```diff
@@ -0,0 +1,297 @@
+ - To include your block type in auto discovery you have to add its id to the `SUPPORTED_THING_TYPES_UIDS` map in `SunSpecConstants`. This is enough for our discovery process to include your thing type in the results.
+
+
+If you have questions or need help don't hesitate to contact us.
```
If you are going to include this line you should probably include contact information as well.

mrbig replied:
Good point made it a bit more clear.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/pom.xml:
```diff
@@ -0,0 +1,31 @@
+    <parent>
+        <groupId>org.openhab.addons.bundles</groupId>
+        <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+        <version>2.5.2-SNAPSHOT</version>
```
```suggestion
        <version>2.5.3-SNAPSHOT</version>
```

mrbig replied:
hmm, is this correct? Maven can not resolve this version at my environment yet. Also I can not find any binding using version 2.5.3-SNAPSHOT at the moment.

mrbig replied:
Thank you, updated to the new version.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecHandlerFactory.java:
```diff
@@ -0,0 +1,89 @@
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<>();
+    static {
+        SUPPORTED_THING_TYPES_UIDS.add(THING_TYPE_INVERTER_SINGLE_PHASE);
+    }
```
Use a Collections.singleton instead.

mrbig replied:
I have a working version that supports several thing types, but removed them from the PR from clarity. Anyways, I replaced this with a singleton for now.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,492 @@
+    protected abstract void handlePolledData(ModbusRegisterArray registers);
+
+    @Override
+    public synchronized void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
```
why is this synchronized?

mrbig replied:
Good point. I was looking up these things from another bundle, but as I see it's not synchronized in the parent class either.
Changed to non-synchronized.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/parser/InverterModelParser.java:
```diff
@@ -0,0 +1,82 @@
+public class InverterModelParser extends AbstractBaseParser implements SunspecParser<InverterModelBlock> {
+
+    @Override
+    public InverterModelBlock parse(ModbusRegisterArray raw) {
```
please remove the blank lines in this method

mrbig replied:
okay, removed those.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/dto/InverterModelBlock.java:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * AC Total Current value
+     */
+    public Integer acCurrentTotal;
```
Why are some of these fields primitive and other not? Aren't the nullable ones supposed to be Optional?

mrbig replied:
Some of the fields are mandatory per the standard specification, while others are optional. The ones that are primitives are the mandatory fields, the Optionals are the optional.

Please note, that these aren't all the attributes that can be parsed from the datasource but there are the lease 3 times as much. Having a constructor with so many parameters would be really ugly imho. However this DTO is written by the parser class that acts like a factory: it makes sure that every field is correctly initialized.

cpmeister replied:
I don't have a problem with the optional ones being represented with an Optional. What I'm trying to understand is what is the difference in this class for int vs Integer and Integer vs Optional<Integer>. I'm trying to figure out where the boxed primitives fit into this.

mrbig replied:
Ok, now I understand your point, and you are right. This was a bit "evolutionary".

I changed everything to classes, so the are also in sync with the optionals.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/resources/ESH-INF/thing/inverter-channel-types.xml:
```diff
@@ -0,0 +1,85 @@
+		<state readOnly="true" pattern="%.1f %unit%"/>
+	</channel-type>
+	
+	<channel-type id="ac-apparent-power-type" advanced="true">
+		<item-type>Number</item-type>
+		<label>AC Apparent Power Value</label>
+		<state readOnly="true" />
+	</channel-type>
+	
+	<channel-type id="ac-reactive-power-type" advanced="true">
+		<item-type>Number</item-type>
+		<label>AC Reactive Power Value</label>
+		<state readOnly="true" />
+	</channel-type>
```
I'd imagine these still have units though right?

mrbig replied:
Yes, I will add them later, but they were missing from openhab-core. So I've added them there first: https://github.com/openhab/openhab-core/pull/1347 While they are merged, I don't see these changes published yet, so I have no way to simply build against those at the moment.

J-N-K replied:
Unfortunately this will be a breaking change later. Even if not physically correct, did you consider using `W` for the time being? This would alt least allow using the correct item-type and not require re-configuration later.

mrbig replied:
Thanks, this is a good idea. I assigned the Watt type, hope this will make the transition easier.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,492 @@
+        if (!endpoint.isPresent()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, String.format(
+                    "Bridge '%s' not completely initialized", Optional.ofNullable(getBridge()).map(b -> b.getLabel())));
+            logger.debug("Bridge not initialized fully (no endpoint) -- aborting init for {}", this);
```
Are you expecting subclasses to implement toString()?

mrbig replied:
Thanks for the comment, I changed this for a more descriptive output

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,492 @@
+    /**
+     * Remove the endpoint if exists
+     */
+    private synchronized void unregisterEndpoint() {
```
You shouldn't synchronize across the handler instance, instead you should use discrete Object instances to do synchronization around.

cpmeister replied:
But now that I've looked further I see no reason for this to synchronized since nothing else is synchronized either. If you want to make sure that `endpoint` is changed in an atomic way I would suggest using `AtomicReference< ModbusSlaveEndpoint>` instead.

mrbig replied:
Thanks for the notice, removed the synchronization.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,492 @@
+            return slaveEndpoint;
+        } else {
+            logger.debug("Unexpected bridge handler: {}", handler);
+            throw new IllegalStateException();
```
Why would this throw an exception rather than return null?

mrbig replied:
Thanks you are now. Changed it for a more coherent behavior.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,493 @@
+    /**
+     * Reference to the modbus manager
+     */
+    protected ModbusManager managerRef;
```
can this be made final?

mrbig replied:
Seems you are right, it can be made final now.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/pom.xml:
```diff
@@ -0,0 +1,31 @@
+    <parent>
+        <groupId>org.openhab.addons.bundles</groupId>
+        <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+        <version>2.5.3-SNAPSHOT</version>
```
We just had a new release so this needs to change again.
```suggestion
        <version>2.5.4-SNAPSHOT</version>
```

mrbig replied:
Thanks for the notice, updated.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,493 @@
+            public void onRegisters(@Nullable ModbusReadRequestBlueprint request,
+                    @Nullable ModbusRegisterArray registers) {
+                if (registers == null) {
+                    logger.info("Received empty register array on poll");
```
does this need to be logged?

mrbig replied:
I would like to keep it. In case there is some error in the communication with the remote device, this could be helpful in debugging the setup.

J-N-K replied:
Agreed. But please reduce to DEBUG

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,493 @@
+    /**
+     * Get a reference to the modbus endpoint
+     */
+    @SuppressWarnings("null")
```
Please don't suppress these warnings, they are usually something you should address.

mrbig replied:
Unfortunately the null checking engine has several false positives. In this method it didn't recognize correctly that the Optional.map() method already handled possible null values.

I moved the SuppressWarnings into the method right before the problematic lines.

If you believe that this is still a potential error, please let me know what is the best practice to fix it.

Thanks.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecHandlerFactory.java:
```diff
@@ -0,0 +1,87 @@
+     */
+    @Activate
+    public SunSpecHandlerFactory(@Reference ModbusManager manager) {
+
```
Removed.

mrbig commented:
> I've picked this apart as much as I can. LGTM

Thank you @cpmeister for your time and effort.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,10 @@
+<features name="org.openhab.binding.modbus.sunspec-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+    <repository>file:${basedirRoot}/bundles/org.openhab.binding.modbus/target/feature/feature.xml</repository>
+
+    <feature name="openhab-binding-modbus-sunspec" description="SunSpec Bundle" version="${project.version}">
```
Should this be installed together with the modbus bundle (like the different bluetooth and mqtt bundles are installed together) or should it be stand-alone?

mrbig replied:
This must be installed with the modbus bundle. It does not work alone at all, because I tried to avoid reimplementing the modbus functionalities.
AFAIK there is no way yet to force the dependency on OH but it would be very welcome here.

J-N-K replied:
I was thinking of the other way: should everyone installing the modbus bundle automatically receive this one? If that‘s the case we would have to change the feature generation.

mrbig replied:
Modbus is a general protocol popular among industrial grade devices and professional home automation.
The basic idea is that we could have a bundle (modbus) that is flexible but needs a decent knowledge from the user, and create higher level bundles for special devices that are more user friendly.
But at the moment this bundle is the only as far as I know.

Pros: bundling this with modbus will make the life of users more simple at the moment
Cons: in the future this _might_ make the modbus bundle pull in lots of unnecessary bundles.

I'm pretty much an outsider in this community, so I won't be able to decide this, so I will leave this to you.

Also I will need help how to change the feature.xml if we decide to install this with the modbus bundle.

J-N-K replied:
We can decide that later and adjust if we decide to bundle them for OH3.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecConstants.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Map of the supported thing type uids, with their block type id
+     */
+    public static final Map<Integer, ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashMap<>();
```
```suggestion
    public static final Map<Integer, ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singletonMap(INVERTER_SINGLE_PHASE, THING_TYPE_INVERTER_SINGLE_PHASE);
```

And let's hope the next entyr comes with OH3 and we can use `Map.of`

mrbig replied:
I've replaced this.
Yes I have implementations ready for 2 other inverter types and 4 meter types ready, but trying to get the basic PR through first.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/dto/InverterModelBlock.java:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Apparent power
+     */
+    public Optional<Short> acApparentPower;
```
Why use an `Optional` here? Since `Optional` is not serializable it shpuld  be avoided as field in DTOs.

mrbig replied:
These classes are not intended to be serialized. They purpose is to server as a middle representation layer between the low level binary types and the high level OH data structures.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,495 @@
+    /**
+     * This is the task used to poll the device
+     */
+    private volatile Optional<PollTask> pollTask = Optional.empty();
```
Please use `@Nullable` here. That's the usual pattern for tasks in all other bindings

mrbig replied:
Ok, I've changed this as you requested.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,495 @@
+            slaveId = slaveEndpointThingHandler.getSlaveId();
+
+            @Nullable
+            ModbusSlaveEndpoint optionalEndpoint = slaveEndpointThingHandler.asSlaveEndpoint();
```
This whole logic seems to be very complicated, compared to a classic null-check.

```
endpoint = slaveEndpointThingHandler.asSlaveEndpoint();
if (endpoint != null) {
   ...
}
```

looks much clearer. If you want to stay with the Optional:

```
  endpoint = Optional.ofNullable(slaveEndpointThingHandler.asSlaveEndpoint());
```

mrbig replied:
Unfortunately your second alternative does not satisfy the eclipse null checker: the Optional.ofNullable has the following signature:
```
Optional<T> Optional.ofNullable(T value)
```
and thus it assumes that T in this cases is also annotated with @NonNull

So far this was the only place where the Optional made me problem compared to the @Nullable approach

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,495 @@
+     * Remove the endpoint if exists
+     */
+    private void unregisterEndpoint() {
+        if (endpoint.isPresent()) {
```
You can skip the `isPresent` check.

mrbig replied:
Thanks, changed this.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/resources/ESH-INF/config/config-descriptions.xml:
```diff
@@ -0,0 +1,39 @@
+
+	<config-description uri="thing-type:sunspec:modbusconfig">
+
+		<parameter name="refresh" type="integer" min="0">
+			<label>Interval (s)</label>
+			<description>Poll interval in seconds. Use zero to disable automatic polling.</description>
+			<default>5</default>
+			<unitLabel>s</unitLabel>
+		</parameter>
```
```suggestion
		<parameter name="refresh" type="integer" min="0" unit="s">
			<label>Interval</label>
			<description>Poll interval. Use zero to disable automatic polling.</description>
			<default>5</default>
		</parameter>
```

mrbig replied:
Thanks, changed.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/SunSpecConfiguration.java:
```diff
@@ -0,0 +1,72 @@
+    /**
+     * Refresh interval in seconds
+     */
+    private long refresh;
```
- add the defaults here
- usually our onfiguration classes have no setters and getters but public fields. is there a special reason why you chose do use another pattern?

mrbig replied:
ok, added a default.

There is no special reason to use getters/setters, so removed them.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,495 @@
+    /**
+     * This is the slave endpoint we're connecting to
+     */
+    protected volatile Optional<ModbusSlaveEndpoint> endpoint = Optional.empty();
```
I think using `@Nullable` instead would simplify your code very much.

mrbig replied:
@J-N-K I disagree here, and it's not just about coding taste.

My problem is with the Eclipse null checker, wich when using @Nullable types gives a dozens of false positives.

More annoyingly it doesn't seem to be coherent: some solutions work at one place, but not at the other. And once I fix everything in the Eclipse IDE and try to compile with maven I get errors again for lines that passed in the IDE or vica versa.

As you requested I've replaced the pollTask with a @Nullable, and now I had to write such ugly hacks:
```java
PollTask task = (@NonNull PollTask) pollTask;
managerRef.unregisterRegularPoll(task);
```
because writing this didn't satisfy the null checker:
```java
if (pollTask != null) {
    managerRef.unregisterRegularPoll(task);
}
```

So after pulling most of my remaining hair I've took a big breath and really replaced everything with @Nonnull Optionals. And while first this seems to be an overkill, during the journey I begun to like it: it's much more talkative, it's easier to recognize where to do the necessary checks, and also it gives a nice fluent api to substitute missing values with defaults.

cpmeister replied:
I wouldn't say that the null checker is wrong (except when it comes to maps) just a bit overzealous. Almost all of the null check warnings can be solved by caching fields to a local variable and then operating on the local variable. The reason that the null checker still complains even after you do a null check is that in a concurrent environment another thread could change your field values as you do your null check. Now normally this isn't the case, but the null checker doesn't know that and it is erring on the side of caution.

So the way that you would solve your example would be to do this:
```
PollTask task = pollTask;
if(task != null) {
    managerRef.unregisterRegularPoll(task);
}
```

mrbig replied:
Yes, I understand this, but then I get the warning from the java compiler, that the given variable can not be null there. Or the IDE keeps removing my lines, because it thinks it's pointless... and this is where I'm starting to pull my hair out.
At the end of day I thinks it's a matter of taste if you like the Optional better or the local variable solution.

Anyway, if this the only thing that blocks my PR being merged, please let me know, and I'll rewrite it.

Thanks for the review.

mrbig commented on bundles/org.openhab.binding.modbus.sunspec/README.md:
```diff
@@ -0,0 +1,297 @@
+| Thing                 | Description                                                           |
+|-----------------------|-----------------------------------------------------------------------|
+| inverter-single-phase | For simple, single phase inverters                                    |
+| inverter-split-phase  | Split phase inverters (Japanese grind and 240V grid in North America) |
```
Thank you, fixed.

cpmeister commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,498 @@
+     * No-op in case no poll task is registered, or if the initialization is incomplete.
+     */
+    private synchronized void unregisterPollTask() {
+        if (pollTask == null) {
+            return;
+        }
+        logger.debug("Unregistering polling from ModbusManager");
+        PollTask task = (@NonNull PollTask) pollTask;
+        managerRef.unregisterRegularPoll(task);
+
+        pollTask = null;
```
This should work, if not let me know and I'll checkout your branch and resolve it myself.
```suggestion
        @Nullable PollTask task = pollTask;
        if(task == null){
            return;
        }
        logger.debug("Unregistering polling from ModbusManager");
        managerRef.unregisterRegularPoll(task);

        pollTask = null;
```

mrbig replied:
Thank you @cpmeister I get it.
Changed the optional attributes in this class to @Nullable types.

kaikreuzer commented:
@J-N-K Are your comments addressed? If so, please approve and we could get it into 2.5.4.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/README.md:
```diff
@@ -0,0 +1,297 @@
+
+| Channel ID           | Item Type | Description                                                         |
+|----------------------|-----------|---------------------------------------------------------------------|
+| ac-total-current     | Number    | Total AC current over all phases in Amperes                         |
```
```suggestion
| ac-total-current     | Number:ElectricCurrent    | Total AC current over all phases in Amperes                         |
```

also adjust the other channels to the correct type

mrbig replied:
ok, adjusted the channel types

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java:
```diff
@@ -0,0 +1,498 @@
+    /**
+     * Configuration instance
+     */
+    @Nullable
+    protected SunSpecConfiguration config = null;
```
```suggestion
    protected @Nullable SunSpecConfiguration config = null;
```

also in other places. And if not needed for a special reason, you can remove the comments above fields. usually the field name is self-explaining its purpose.

mrbig replied:
Ok, changed how @Nullable is written.
Regarding the comments: a few more words in the comments might help someone in the future IMHO.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/resources/ESH-INF/config/config-descriptions.xml:
```diff
@@ -0,0 +1,38 @@
+
+	<config-description uri="thing-type:sunspec:modbusconfig">
+
+		<parameter name="refresh" type="integer" min="0">
+			<label>Interval</label>
+			<description>Poll interval in seconds. Use zero to disable automatic polling.</description>
+			<default>5</default>
+		</parameter>
```
```suggestion
		<parameter name="refresh" type="integer" min="0" unit="s">
			<label>Refresh Interval</label>
			<description>Poll interval. Use zero to disable automatic polling.</description>
			<default>5</default>
		</parameter>
```

mrbig replied:
Thanks, changed this.

J-N-K commented on bundles/org.openhab.binding.modbus.sunspec/src/main/resources/ESH-INF/thing/inverter-types.xml:
```diff
@@ -0,0 +1,36 @@
+		</supported-bridge-type-refs>
+
+		<label>Single Phase Inverter</label>
+		<description>Single phase inverter supporting SunSpec mapping over tcp modbus connection</description>
```
```suggestion
		<description>Single phase inverter supporting SunSpec mapping over tcp modbus connection.</description>
```

mrbig replied:
Thanks, fixed.

ssalonen commented:
It would be great to see this in 2.5.4! Do we need new core release as well to support the UoM changes required by this PR?

https://github.com/openhab/openhab-addons/pull/6331#issuecomment-580748051

mrbig commented:
I'm very happy that this could make it to the mainline. Thank you guys for reviewing my work and all the help and feedback I've got from you. Specially for @J-N-K @cpmeister @Hilbrand @ssalonen 

Secondly there's PR #6994 wich is is closely related follow up to this binding. It adds auto discovery to this bundle and that makes it really user friendly. I hope it can make it to the mainline in near future as well, but I know everyone is busy these days.

Thanks for all the work!

cweitkamp commented:
@ssalonen

> Do we need new core release as well to support the UoM changes required by this PR?

The new units will be available for add-ons after a new OHC release. But we do not plan to release a 2.5.x core anymore. You will have to wait for OH3 to adapt the related lines of code.

ssalonen commented:
@cweitkamp I'm not sure what will be the implications in practice for sunspec binding -- it was included in openHAB 2.5.4 (addons?): https://github.com/openhab/openhab-distro/releases/tag/2.5.4

cweitkamp commented:
@ssalonen There won't be any. @mrbig used `SmartHomeUnits.WATT` right now and left a TODO comment in the code to resolve it later.
